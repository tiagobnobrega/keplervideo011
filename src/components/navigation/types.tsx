/*
 * Copyright (c) 2024 Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * PROPRIETARY/CONFIDENTIAL.  USE IS SUBJECT TO LICENSE TERMS.
 */

import { DrawerScreenProps } from '@amzn/react-navigation__drawer/lib/typescript/src/types';
import { StackNavigationProp } from '@amzn/react-navigation__stack';
import { StackScreenProps } from '@amzn/react-navigation__stack/lib/typescript/src/types';
import { ImageSourcePropType, TouchableOpacity } from 'react-native';
import { TitleData } from '../../types/TitleData';

export enum Screens {
  DEFAULT_SCREEN = 'Home',
  APP_DRAWER = 'AppDrawer',
  HOME_SCREEN = 'Home',
  DETAILS_SCREEN = 'Details',
  PLAYER_SCREEN = 'Player',
  SETTINGS_SCREEN = 'Settings',
  SEARCH_SCREEN = 'Search',
  SEARCH_RESULTS_SCREEN = 'SearchResultsScreen',
  FEEDBACK_SCREEN = 'FeedBackScreen',
}

export enum DrawerType {
  PERMANENT = 'permanent',
  FRONT = 'front',
  SLIDE = 'slide',
  BACK = 'back',
}

export type AppStackParamList = {
  [Screens.APP_DRAWER]: undefined;
  [Screens.HOME_SCREEN]: undefined;
  [Screens.SEARCH_SCREEN]: undefined;
  [Screens.SETTINGS_SCREEN]: undefined;
  [Screens.FEEDBACK_SCREEN]: undefined;
  [Screens.SEARCH_RESULTS_SCREEN]: {
    searchKeyword: string;
  };
  [Screens.DETAILS_SCREEN]: {
    data: TitleData;
    sendDataOnBack: (id: number | string) => void;
  };
  [Screens.PLAYER_SCREEN]: {
    data: TitleData;
    sendDataOnBack: () => void;
  };
};

export type AppStackScreenProps<T extends keyof AppStackParamList> =
  StackScreenProps<AppStackParamList, T>;

export type HomeScreenNavigationProps = StackNavigationProp<
  AppStackParamList,
  Screens.HOME_SCREEN
>;

export type AppDrawerParamList = {
  [Screens.HOME_SCREEN]: undefined;
  [Screens.SETTINGS_SCREEN]: undefined;
  [Screens.SEARCH_SCREEN]: undefined;
};

export type AppDrawerScreenProps<T extends keyof AppDrawerParamList> =
  DrawerScreenProps<AppDrawerParamList, T>;

declare global {
  namespace ReactNavigation {
    interface RootParamList extends AppStackParamList {}
  }
}

export interface ButtonConfig {
  onPress: () => void;
  image: ImageSourcePropType;
  label: string;
  ref?: React.RefObject<TouchableOpacity>;
  testID: string;
}
