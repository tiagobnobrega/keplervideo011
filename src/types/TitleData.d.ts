import { MediaItem } from '@amzn/keplerblocks-types';

export type MimeType =
  | 'text/plain'
  | 'text/vtt'
  | 'application/x-subtitle' // .srt
  | 'application/x-subtitle-vtt' // .vtt
  | 'application/x-subtitle-sami'
  | 'application/x-subtitle-tmplayer'
  | 'application/x-subtitle-mpl2'
  | 'application/x-subtitle-dks'
  | 'application/x-subtitle-qttext'
  | 'application/x-subtitle-lrc';

export type VideoFormat = 'MPD' | 'HLS' | 'MP4' | 'DASH';

interface DrmScheme {
  name: string;
  licenseUri: string;
  headerTag?: string;
  headerData?: string;
}
interface TextTrack {
  label: string;
  language: string;
  uri: string;
  mimeType: MimeType;
}

export interface TitleData extends MediaItem {
  videoUrl: string;
  categories: string[];
  channelID: string;
  posterUrl: string;
  format: string;
  drmScheme?: DrmScheme;
  textTrack?: TextTrack[];
  uhd: boolean;
  secure: boolean;
  aCodec?: string;
  vCodec?: string;
  rentAmount: string;
}
