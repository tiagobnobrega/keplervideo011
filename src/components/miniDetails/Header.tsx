import { isEqual } from 'lodash';
import React from 'react';
import { StyleSheet, Text, View } from 'react-native';
import { COLORS } from '../../styles/Colors';
import { TitleData } from '../../types/TitleData';
import { scaleUxToDp } from '../../utils/pixelUtils';

interface HeaderProps {
  data?: TitleData;
}

const Header = ({ data }: HeaderProps) => {
  return (
    <View style={styles.headerContent}>
      <Text style={styles.titleText} numberOfLines={2}>
        {data?.title}
      </Text>
      <Text style={styles.descriptionText} numberOfLines={3}>
        {data?.description}
      </Text>
    </View>
  );
};

const headerDataUnchanged = (
  prevProps: HeaderProps,
  nextProps: HeaderProps,
) => {
  return isEqual(prevProps.data, nextProps.data);
};

export default React.memo(Header, headerDataUnchanged);

const styles = StyleSheet.create({
  headerContent: {
    paddingStart: scaleUxToDp(120),
    paddingTop: scaleUxToDp(138),
    width: scaleUxToDp(884),
  },
  titleText: {
    fontSize: scaleUxToDp(60),
    fontWeight: 'bold',
    color: COLORS.SMOKE_WHITE,
    width: scaleUxToDp(800),
  },
  descriptionText: {
    fontSize: scaleUxToDp(30),
    color: COLORS.SMOKE_WHITE,
    width: scaleUxToDp(884),
  },
});
