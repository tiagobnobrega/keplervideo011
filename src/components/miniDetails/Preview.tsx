import { isEqual } from 'lodash';
import React from 'react';
import { Image, StyleSheet, View } from 'react-native';
import { scaleUxToDp } from '../../utils/pixelUtils';
const LeftGradient = require('../../assets/mini_details_left_gradient.png');
const BottomGradient = require('../../assets/mini_details_bottom_gradient.png');

interface PreviewProps {
  posterUrl?: string;
}

const Preview = ({ posterUrl }: PreviewProps) => {
  return (
    <View style={styles.previewContent}>
      {posterUrl && (
        <Image
          testID={'poster-image'}
          style={styles.posterImage}
          source={{ uri: posterUrl }}
          resizeMode={'contain'}
        />
      )}
      <GradientOverlay />
    </View>
  );
};

const GradientOverlay = () => {
  return (
    <View>
      <Image
        style={styles.gradient}
        resizeMode={'contain'}
        source={LeftGradient}
        testID="left-gradient"
      />
      <Image
        style={styles.gradient}
        resizeMode={'contain'}
        source={BottomGradient}
        testID="bottom-gradient"
      />
    </View>
  );
};

const posterUrlDataUnchanged = (
  prevProps: PreviewProps,
  nextProps: PreviewProps,
) => {
  return isEqual(prevProps.posterUrl, nextProps.posterUrl);
};

export default React.memo(Preview, posterUrlDataUnchanged);

const styles = StyleSheet.create({
  gradient: {
    width: scaleUxToDp(1236),
    height: scaleUxToDp(696),
    position: 'absolute',
  },
  previewContent: {
    left: scaleUxToDp(682),
    position: 'absolute',
  },
  posterImage: {
    width: scaleUxToDp(1236),
    height: scaleUxToDp(696),
    position: 'absolute',
  },
});
