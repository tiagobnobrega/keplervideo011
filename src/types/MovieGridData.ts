import { TitleData } from './TitleData';

export type MovieGridData = {
  heading: string;
  testID: string;
  data: () => TitleData[];
};
