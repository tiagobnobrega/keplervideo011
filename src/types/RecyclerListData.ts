import { StyleProp, ViewStyle } from 'react-native';
import { RecyclerListViewProps } from 'recyclerlistview';

export interface CustomRecyclerListViewProps extends RecyclerListViewProps {
  contentContainerStyle?: StyleProp<ViewStyle>;
  initialNumToRender: number | undefined;
}
