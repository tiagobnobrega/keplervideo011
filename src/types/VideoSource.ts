// adapted from GStreamer MIME Types (https://gstreamer.freedesktop.org/documentation/subparse/subparse.html?gi-language=c)
export type MimeType =
  | 'text/plain'
  | 'text/vtt'
  | 'application/x-subtitle' // .srt
  | 'application/x-subtitle-vtt' // .vtt
  | 'application/x-subtitle-sami'
  | 'application/x-subtitle-tmplayer'
  | 'application/x-subtitle-mpl2'
  | 'application/x-subtitle-dks'
  | 'application/x-subtitle-qttext'
  | 'application/x-subtitle-lrc';

export type VideoFormat = 'MPD' | 'HLS' | 'MP4' | 'DASH';

interface DrmScheme {
  name: string;
  licenseUri: string;
  headerTag?: string;
  headerData?: string;
}
interface TextTrack {
  label: string;
  language: string;
  uri: string;
  mimeType: MimeType;
}

export interface VideoSource {
  title: string;
  uri: string;
  format: VideoFormat;
  drmScheme?: DrmScheme;
  textTrack?: TextTrack[];
  uhd: boolean;
  secure: boolean;
  aCodec?: string;
  vCodec?: string;
}
