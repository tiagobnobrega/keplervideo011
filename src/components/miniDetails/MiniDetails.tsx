import { isEqual } from 'lodash';
import React from 'react';
import { StyleSheet, View } from 'react-native';
import { TitleData } from '../../types/TitleData';
import Header from './Header';
import Preview from './Preview';

interface MiniDetailsProps {
  selectedTitle?: TitleData;
}

const MiniDetails = ({ selectedTitle }: MiniDetailsProps) => {
  return (
    <View style={styles.container}>
      <Preview posterUrl={selectedTitle?.thumbnail} />
      <Header data={selectedTitle} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    flex: 1,
  },
});

const selectedTitleUnchanged = (
  prevProps: MiniDetailsProps,
  nextProps: MiniDetailsProps,
) => {
  return isEqual(prevProps.selectedTitle, nextProps.selectedTitle);
};

export default React.memo(MiniDetails, selectedTitleUnchanged);
