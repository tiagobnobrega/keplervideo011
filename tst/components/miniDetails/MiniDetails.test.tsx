import { render } from '@testing-library/react-native';
import { isEqual } from 'lodash';
import React from 'react';
import 'react-native';
import Header from '../../../src/components/miniDetails/Header';
import MiniDetails from '../../../src/components/miniDetails/MiniDetails';
import Preview from '../../../src/components/miniDetails/Preview';
import { TitleData } from '../../../src/types/TitleData';

jest.mock('lodash', () => ({
  isEqual: jest.fn(),
}));

const mockSelectedTitle: TitleData = {
  id: '1',
  title: 'Sample Title',
  thumbnail: 'http://example.com/thumbnail.jpg',
  videoUrl: '',
  categories: [],
  channelID: '',
  posterUrl: '',
  format: '',
  description: 'Sample Description',
  mediaType: 'video',
  mediaSourceType: 'url',
  uhd: false,
  secure: false,
  rentAmount: '',
};

describe('Testing MiniDetails component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    (jest.mocked(isEqual) as jest.Mock).mockImplementation((a, b) => {
      return JSON.stringify(a) === JSON.stringify(b);
    });
  });

  it('renders correctly and matches snapshot', () => {
    const { toJSON } = render(
      <MiniDetails selectedTitle={mockSelectedTitle} />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders Preview and Header components', () => {
    const { queryAllByText } = render(
      <MiniDetails selectedTitle={mockSelectedTitle} />,
    );
    const headerComponents = queryAllByText(mockSelectedTitle.title);
    expect(headerComponents.length).toBeLessThanOrEqual(0);
  });

  it('renders correctly with components', () => {
    render(<MiniDetails selectedTitle={mockSelectedTitle} />);
    const headerText = mockSelectedTitle.title;
    expect(headerText).toBeTruthy();
  });

  it('renders with components', () => {
    render(<MiniDetails selectedTitle={mockSelectedTitle} />);
    expect(Preview).toBeTruthy();
    expect(Header).toBeTruthy();
  });

  it('does not re-render when selectedTitle is unchanged', () => {
    const { rerender } = render(
      <MiniDetails selectedTitle={mockSelectedTitle} />,
    );
    rerender(<MiniDetails selectedTitle={mockSelectedTitle} />);
    expect(isEqual).toHaveBeenCalledTimes(0);
  });

  it('re-renders when selectedTitle changes', () => {
    const { rerender } = render(
      <MiniDetails selectedTitle={mockSelectedTitle} />,
    );
    const newSelectedTitle: TitleData = {
      id: '2',
      title: 'New Title',
      thumbnail: 'http://example.com/new-thumbnail.jpg',
      videoUrl: '',
      categories: [],
      channelID: '',
      posterUrl: '',
      format: '',
      description: 'New Description',
      mediaType: 'video',
      mediaSourceType: 'url',
      uhd: false,
      secure: false,
      rentAmount: '',
    };
    rerender(<MiniDetails selectedTitle={newSelectedTitle} />);
    expect(isEqual).not.toHaveBeenCalled();
  });
});
