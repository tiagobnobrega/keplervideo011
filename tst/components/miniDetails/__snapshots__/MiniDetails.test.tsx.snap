// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing MiniDetails component renders correctly and matches snapshot 1`] = `
<MiniDetails
  selectedTitle={
    Object {
      "categories": Array [],
      "channelID": "",
      "description": "Sample Description",
      "format": "",
      "id": "1",
      "mediaSourceType": "url",
      "mediaType": "video",
      "posterUrl": "",
      "rentAmount": "",
      "secure": false,
      "thumbnail": "http://example.com/thumbnail.jpg",
      "title": "Sample Title",
      "uhd": false,
      "videoUrl": "",
    }
  }
/>
`;
